### PR TITLE
Js Scope - Coding Logic

### LINKED ISSUE #197: Js Scope - Coding - Logic

### LINKED ISSUE #217: Substitute - Overload API
Revisit this "hack", we'd want to improve the API for substitute so that it's able to take in just values (array) instead of objects, while maintaining all other current use cases

_Originally posted by @francinelucca in https://github.com/ibm-telemetry/telemetry-js/pull/198#discussion_r1526771566_

### LINKED ISSUE #220: Js Scope - Unit Testing

### PR DESCRIPTION
**Pull request by francinelucca**
Closes #197, Closes #220, Closes #217

- Adds the programming logic that detects tokens and functions and maps them to the instrumented package, ultimately generating token and function metrics.
- Adds unit testing for all things Js Scope.
- Reworks substitution logic to be able to take in arrays and objects.
- `createSourceFileFromText` function for unit testing ease-of-use
- Fixes significant bug in `getPackageData` that was preventing collection under certain circumstances.

#### Changelog

**New**

- Split substitution logic into new`substituteArray` and `substituteObject` functions, along with a higher-logic `Substitution` class.
- New `getAccessPath` function determines the access path to a function or token (e.g. ['object', 'property', 'token'].
- `deduplicateFunctions` and `deduplicatedTokens` inner functions in `JsScope`.
- New project and sample fixtures for JS scope unit testing.
- Test files containing unit tests and e2e scope for added functions and logic.
- "empty-package-json" project fixture to test edge case.
- Adds test file and cases for `findRelevantSourceFiles`

**Changed**

- Updates `@ibm/telemetry-attributes-js` version to `3.1.0`
- Refines exit condition in `getPackageData` to ignore empty packages.
- Excludes test fixtures from sonarcloud reporting.
- Makes `ComplexValue` readonly.
- Consolidates function-import and token-import matchers to be a combined set of "js-*-import-matcher"s
- Fixes overcollection bug on `JsNamedImportMatcher` & `JsRenamedImportMatcher`
- Consolidates `PropertyAccessExpressionHandler` & `ElementAccessExpressionNodeHandler` to single `AccessExpressionNodeHandler`
- Add `startPos` and `endPos` to `JsToken` & `JsFunction` interfaces (used to track token and function duplication)
- Allow `accessPath` property of `JsToken` & `JsFunction` interfaces to take in `ComplexValue`s (used for sensitive information substitution)
- Remove `readonly` property of `JsAccumulator`'s `imports` (gets rewrittein in `removeIrrelevantImports`)
- Remove `readonly` property of `JsFunctionTokenAccumulator`'s `tokens` and `functions` (gets rewrittein in deduplicate functions)
- `FunctionMetric`:
     - Use new substitution logic
     - Add substitution logic to handle renames, `all` imports and complex values.
     - Add `FUNCTION_MODULE_SPECIFIER`
     - Returns `FUNCTION_ARGUMENT_VALUES` as a space-separated string instead of array.
- `TokenMetric`:
     - Use new substitution logic
     - Add substitution logic to handle renames, `all` imports and complex values.
     - Add `TOKEN_MODULE_SPECIFIER`
- Implement `CallExpressionNodeHandler` logic to catch and build `JsFunction`s
- Implements `IdentifierNodeHandler` logic to catch and build `JsToken`s from identifiers
- Use `substituteObject` function in `ElementMetric`

**Removed**

- src/main/core/anonymize/substitute in favor of `substituteObject` and `substituteArray` functions.

#### Testing / reviewing

All unit tests should pass, test coverage should be above threshold and logic should read seamlessly.

Local Test:
- Pull down this branch
- `npm install`
- `npm run build`
- Go to a local clone of a repository that contains an instrumented package (say [carbon-website](https://github.com/carbon-design-system/carbon-website) or [carbon-platform](https://github.com/carbon-design-system/carbon-platform))
- Update local `telemetry.yml` to include js scope: 
```
collect:
...
  js:
    functions: {}
    tokens:
```
- From inside the instrumented package's directory, run `CI=true node /path-to-telemetry/telemetry-js/dist/collect.js --config=telemetry.yml`, verify log file

### CODE REVIEW

File: src/main/scopes/js/get-access-path.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import { getNodeValueHandler } from './node-value-handler-map.js'
+
+/**
+ * Constructs a JsToken object from a given Identifier type AST node.
+ *
+ * @param node - TS node to retrieve access path for.
+ * @param sourceFile - Root AST node.
+ * @param logger - A logger instance.
+ * @param currAccessPath - For internal use only, tracks current constructed access path.
+ * @param topLevel - For internal use only, tracks top level function call.
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/get-access-path.ts:20]: As a convention, since none of these params are optional, let's make `logger` be the last one in the function's params.

File: src/main/scopes/js/get-access-path.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * @returns Access path to node represented as string array (chunks).
+ */
+export default function getAccessPath(
+  node: ts.Node,
+  sourceFile: ts.SourceFile,
+  logger: Logger,
+  currAccessPath: Array<string | ComplexValue> = [],
+  topLevel: boolean = true
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/get-access-path.ts:28]: To hide this from the API, I'd recommend creating two functions, one that's the private, internal, recursive one; and another that's the public API without `curAccessPath` or `topLevel`

File: src/main/scopes/js/get-access-path.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * LICENSE file in the root directory of this source tree.
+ */
+import * as ts from 'typescript'
+
+import type { Logger } from '../../core/log/logger.js'
+import { ComplexValue } from './complex-value.js'
+import { getNodeValueHandler } from './node-value-handler-map.js'
+
+/**
+ * Constructs a JsToken object from a given Identifier type AST node.
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/get-access-path.ts:14]: Since this is a fairly complex function, it might be worth replacing this with a longer explanation of what this function does and how it accomplishes it, and if there are any restrictions or edge cases that should be known to the caller.

File: src/main/scopes/js/import-matchers/js-all-import-matcher.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import type { JsFunction, JsImport, JsImportMatcher, JsToken } from '../interfaces.js'
+
+/**
+ * Identifies JsTokens or JsFunctions that have been imported as all imports,
+ * and returns an import element match (if any) or undefined otherwise.
+ */
+export class JsAllImportMatcher implements JsImportMatcher<JsToken | JsFunction> {
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/import-matchers/js-all-import-matcher.ts:14]: For these classes that basically just implement an interface, you can keep these class-level docs pretty basic, like "Import matcher for all (*) imports, such as `import * as stuff from 'whatever'`" since it probably makes more sense to put the bulk of the explanation on the method below instead (like you have).

File: src/main/scopes/js/import-matchers/js-named-import-matcher.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import type { JsFunction, JsImport, JsImportMatcher, JsToken } from '../interfaces.js'
+
+/**
+ * Identifies JsTokens and JsFunctions that have been imported as named imports,
+ * and returns an import element match (if any) or undefined otherwise.
+ */
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/import-matchers/js-named-import-matcher.ts:13]: Same thing here. You can consolidate these docs on the `findMatch` method and keep the class level docs pretty basic.

File: src/main/scopes/js/import-matchers/js-all-import-matcher.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+   * (.e.g: import * as something from 'package')
+   * that matches the supplied list of import elements.
+   *
+   * @param jsElement - JsToken or JsFunction to evaluate.
+   * @param imports - Import elements to use for comparison.
+   * @returns Corresponding JsImport element if token was imported as an all import,
+   * undefined otherwise.
+   */
+  findMatch(jsElement: JsToken | JsFunction, imports: JsImport[]) {
+    return jsElement.accessPath.length >= 2
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/import-matchers/js-all-import-matcher.ts:26]: It might be good to document why `2` is special here either by making it a named constant or adding a comment above this line.

File: src/main/scopes/js/import-matchers/js-named-import-matcher.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+   * (e.g.: `import {something} from 'package'`).
+   *
+   * @param jsElement - JsToken or JsFunction to evaluate.
+   * @param imports - Import elements to use for comparison.
+   * @returns Corresponding JsImport if token was imported as a name import,
+   * undefined otherwise.
+   */
+  findMatch(jsElement: JsToken | JsFunction, imports: JsImport[]) {
+    return imports.find(
+      (i) => !i.isDefault && !i.isAll && !Boolean(i.rename) && i.name === jsElement.accessPath[0]
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/import-matchers/js-named-import-matcher.ts:26]:
```suggestion
      (i) => !i.isDefault && !i.isAll && i.rename === undefined && i.name === jsElement.accessPath[0]
```

File: src/main/scopes/js/import-matchers/js-renamed-import-matcher.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+   * (e.g.: `import {something as somethingElse} from 'package'`)
+   * and returns an import element match (if any) or undefined otherwise.
+   *
+   * @param jsElement - JsToken or JsFunction to evaluate.
+   * @param imports - Import elements to use for comparison.
+   * @returns Corresponding JsImport if token was imported as a renamed import,
+   * undefined otherwise.
+   */
+  findMatch(jsElement: JsToken | JsFunction, imports: JsImport[]) {
+    return imports.find((i) => Boolean(i.rename) && i.rename === jsElement.accessPath[0])
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/import-matchers/js-renamed-import-matcher.ts:25]:
```suggestion
    return imports.find((i) => i.rename !== undefined && i.rename === jsElement.accessPath[0])
```

File: src/main/scopes/js/metrics/function-metric.ts [hunk 9]
```diff
@@ -71,24 +74,48 @@ export class FunctionMetric extends ScopeMetric {
...
+      functionName = functionName.replace(this.matchingImport.rename, this.matchingImport.name)
+      // replace the import name in access path
+      functionAccessPath[0] = this.matchingImport.name
+    }
+
+    const subs = new Substitution()
+    // redact complex values
+    functionAccessPath.forEach((segment) => {
+      if (segment instanceof ComplexValue) {
+        functionName = safeStringify(functionName).replace(
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/metrics/function-metric.ts:91]: minor: This `safeStringify` call can be moved up to the let statement on line 77 so that it doesn't need to be re-called every iteration.

File: src/main/scopes/js/metrics/function-metric.ts [hunk 10]
```diff
@@ -71,24 +74,48 @@ export class FunctionMetric extends ScopeMetric {
...
+          safeStringify(segment.complexValue),
+          subs.put(segment)
+        )
+      }
+    })
+
+    // redact "all" import
+    if (this.matchingImport.isAll) {
+      functionAccessPath[0] = subs.put(this.matchingImport.name)
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/metrics/function-metric.ts:100]: This unfortunately still mutates the field on the class 😬 
Since `functionAccessPath` is a reference to `this.jsFunction.accessPath`, modifying its elements will modify the original array.

File: src/main/scopes/js/metrics/function-metric.ts [hunk 11]
```diff
@@ -71,24 +74,48 @@ export class FunctionMetric extends ScopeMetric {
...
-      [JsScopeAttributes.FUNCTION_ACCESS_PATH]: this.jsFunction.accessPath,
-      [JsScopeAttributes.FUNCTION_ARGUMENT_VALUES]: Object.values(anonymizedArguments).map((arg) =>
-        String(arg)
-      ),
+      [JsScopeAttributes.FUNCTION_NAME]: functionName,
+      // TODO: uncomment
+      // [JsScopeAttributes.FUNCTION_MODULE_SPECIFIER]: this.matchingImport.path,
+      [JsScopeAttributes.FUNCTION_ACCESS_PATH]: substituteArray(
+        functionAccessPath,
+        functionAccessPath.filter((p) => typeof p === 'string')
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/metrics/function-metric.ts:112]: Am I correct in understanding that what this is doing is to allow any non-complex value from the access path under the assumption that for this type of metric, the access path would have to be entirely from sub-parts of an instrumented function?

File: src/test/__fixtures/js-samples/all-js-function-types.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+anObject['property'].aFunction()
+
+anObject['property'][4]('hey', 500, {yeah: 'yeah'})
+
+anObject[BLA['property']]['aFunction']({variable, ...object})
+
+anObject.property.aFunction?.('click', aFunction)
+
+aFunction().anotherFunction().anotherFunction()
```
Comment (jdharvey-ibm) [inline src/test/__fixtures/js-samples/all-js-function-types.ts:23]:
```suggestion
aFunction().anotherFunction().anotherFunction()

```

File: src/test/__fixtures/js-samples/all-js-token-types.ts [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  
+blue['60'][THE_CONSTANT]?.['bloop']
+  
+const bla = THE_CONSTANT
+
+anObject[BLA['property']]['anotherProperty']
+
+anObject[BLA.property]['anotherProperty']
+
+BLA['property'][A_CONSTANT]
```
Comment (jdharvey-ibm) [inline src/test/__fixtures/js-samples/all-js-token-types.ts:27]:
```suggestion
BLA['property'][A_CONSTANT]

```

File: src/test/__fixtures/projects/basic-project/test.js [hunk 14]
```diff
@@ -1,2 +1,9 @@
 
 // @ts-nocheck
+import * as BLA from 'instrumented/the/path'
+
+const A = BLA['hi']
+
+BLA.someFunction(1,2,3, [4,5,6], true, false, 32, {an: 'object'})
+
+one.two.three['four']
```
Comment (jdharvey-ibm) [inline src/test/__fixtures/projects/basic-project/test.js:9]:
```suggestion
one.two.three['four']

```

File: src/test/__fixtures/projects/basic-project/test.jsx [hunk 15]
```diff
@@ -8,3 +9,7 @@ export {
   sample,
   secondSample
 }
+
+anObject[BLE.property]['anotherProperty']
+
+BLE.property.aFunction?.('click', aFunction)
```
Comment (jdharvey-ibm) [inline src/test/__fixtures/projects/basic-project/test.jsx:15]:
```suggestion
BLE.property.aFunction?.('click', aFunction)

```

File: src/test/__fixtures/projects/basic-project/test.ts [hunk 16]
```diff
@@ -4,6 +4,11 @@
...
  * LICENSE file in the root directory of this source tree.
  */
+// @ts-nocheck
+import * as BLA from 'not/instrumented'
+
 export function unrelated() {
   return 'hello world'
 }
+
+BLA.functionCall()
```
Comment (jdharvey-ibm) [inline src/test/__fixtures/projects/basic-project/test.ts:14]:
```suggestion
BLA.functionCall()

```

There's a setting you can flip on in vscode to automatically add a new line to the end of files on save. I definitely recommend it as it gets rid of these "no newline" red warning icons at the ends of files.

File: src/test/__fixtures/projects/basic-project/test.tsx [hunk 17]
```diff
@@ -27,3 +27,7 @@ export const MyComponent = ({other, ...spreadObj}) => {
     </OtherThing>
   )
 }
+
+anObject['property'][A_RENAMED_TOKEN]('hey', 500, {yeah: 'yeah'})
+
+Function1()
```
Comment (jdharvey-ibm) [inline src/test/__fixtures/projects/basic-project/test.tsx:33]:
```suggestion
Function1()

```

File: src/test/scopes/js/import-matchers/js-named-import-matcher.test.ts [hunk 18]
```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright IBM Corp. 2023, 2024
```
Comment (jdharvey-ibm) [inline src/test/scopes/js/import-matchers/js-named-import-matcher.test.ts:2]: For these net-new test files, make sure the date at the top is `2024, 2024`. Unless they're refactors of the jsx stuff. It's hard to tell from here since git isn't treating it as a rename if it is one for a lot of these files.

File: src/main/scopes/js/metrics/token-metric.ts [hunk 19]
```diff
@@ -63,9 +67,40 @@ export class TokenMetric extends ScopeMetric {
...
     )
 
+    let tokenName = this.jsToken.name
+    const tokenAccessPath = this.jsToken.accessPath
+
+    // Handle renamed tokens
+    if (this.matchingImport.rename !== undefined) {
+      tokenName = tokenName.replace(this.matchingImport.rename, this.matchingImport.name)
+      // replace import name in access path
+      tokenAccessPath[0] = this.matchingImport.name
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/metrics/token-metric.ts:77]: Same comment here and on line 93 about mutating the input instance variables.

File: src/main/scopes/js/metrics/token-metric.ts [hunk 20]
```diff
@@ -63,9 +67,40 @@ export class TokenMetric extends ScopeMetric {
...
+    }
+
+    const subs = new Substitution()
+    // redact complex values
+    tokenAccessPath.forEach((segment) => {
+      if (segment instanceof ComplexValue) {
+        tokenName = safeStringify(tokenName).replace(
```
Comment (jdharvey-ibm) [inline src/main/scopes/js/metrics/token-metric.ts:84]: minor: Same comment here about being able to move the safeStringify call up to the assignment statement.

File: src/main/scopes/js/metrics/function-metric.ts (hunk 11 above)
Comment (francinelucca) [reply src/main/scopes/js/metrics/function-metric.ts:109]: Correct, the accessPath of a JsFunction can only be a string or a ComplexValue, as per defined by the [interface](https://github.com/ibm-telemetry/telemetry-js/pull/216/files#diff-57923b99fcb2d80f0d95a95d5f5b0adc4a1c74c69ab551abf8dcf9a4e1ab2f5cR49), so i'm just allowing all strings so all complex values get substituted.

**Review by jdharvey-ibm — APPROVED** [review]
